### PR TITLE
nixos cups: support for socket activation [WIP]

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

With printing enabled, the cups services would start eagerly on boot despite only being needed when printing.

With this PR, cups only start when needed so it takes a few seconds for the print dialog to show.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by Mic92 — APPROVED** [review]
I have not tested the pull request, but I agree with the idea.

File: nixos/modules/services/printing/cupsd.nix [hunk 1]
```diff
@@ -259,8 +280,31 @@ in
           CUPSd temporary directory.
         '';
       };
-    };
 
+      startWhenNeeded = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Start CUPS processes on demand.
```
Comment (coretemp) [inline nixos/modules/services/printing/cupsd.nix]: Specifying expected extra latency on a specific piece of hardware would be useful. Something like 

`On my Intel-<specific model> this delays a printing dialog by 2.4 seconds`.

File: nixos/modules/services/printing/cupsd.nix [hunk 2]
```diff
@@ -285,73 +329,85 @@ in
...
-            ${optionalString (containsGutenprint cfg.drivers) ''
-              if [ -d /var/lib/cups/ppd ]; then
-                ${getGutenprint cfg.drivers}/bin/cups-genppdupdate -p /var/lib/cups/ppd
-              fi
-            ''}
-          '';
+          fi
+          # Then, populate it with static files
+          cd ${rootdir}/etc/cups
+          for i in *; do
```
Comment (coretemp) [inline nixos/modules/services/printing/cupsd.nix]: `*` needs to be `./*`.

File: nixos/tests/printing.nix [hunk 3]
```diff
@@ -35,12 +35,11 @@ import ./make-test.nix ({pkgs, ... }: {
       startAll;
 
       # Make sure that cups is up on both sides.
-      $server->waitForUnit("cups.service");
-      $client->waitForUnit("cups.service");
-      $client->sleep(10); # wait until cups is fully initialized
+      $server->waitForUnit("sockets.target");
```
[the remark this answers is not in the document]
Comment (peterhoeg) [reply nixos/tests/printing.nix]: Indeed they should - that is one of the 2 issues I want to clean up before this gets merged.

**Comment by mmahut** [comment]
Are there any updates on this pull request, please?

**Pull request closed without merging**